datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner_test.py

```diff
@@ -0,0 +1,270 @@
+            ["drop-reminder", 10, 7, 2],
+            ["strict", 12, 6, 3],
+            ["allow-smaller", 12, 6, 3],
+            ["allow-bigger", 12, 6, 3],
+            ["drop-reminder", 12, 6, 3],
+        ]
+    )
+    def test_no_overlapping_natural_ids(
+        self,
+        mode: Literal["allow-smaller", "allow-bigger", "drop-reminder", "strict"],
+        num_rows: int,
+        num_unique_natural_id: int,
+        group_size: int,
+    ) -> None:
+        """Test that no natural_ids overlap across partitions."""
+        dataset = _create_dataset(num_rows, num_unique_natural_id)
+        partitioner = GroupedNaturalIdPartitioner(
+            partition_by="natural_id",
+            group_size=group_size,
+            mode=mode,
+            sort_unique_ids=self.sort_unique_ids,
+        )
+        partitioner.dataset = dataset
+
+        # Trigger partitioning
+        partitions = [
+            partitioner.load_partition(i) for i in range(partitioner.num_partitions)
+        ]
+
+        # Check for overlaps between partitions
+        seen_natural_ids: Set[str] = set()
+        for partition in partitions:
+            natural_ids_in_partition = set(partition.unique("natural_id"))
+
+            # Check if there is any overlap with previously seen natural IDs
+            overlap = seen_natural_ids.intersection(natural_ids_in_partition)
+            self.assertTrue(
+                len(overlap) == 0,
+                f"Overlapping natural IDs found between partitions in mode: {mode}. "
+                f"Overlapping IDs: {overlap}",
+            )
+
+            # Add the natural IDs from this partition to the seen set
+            seen_natural_ids.update(natural_ids_in_partition)
+
+    def test_strict_mode_with_invalid_group_size(self) -> None:
+        """Test strict mode raises if group_size does not divide unique IDs evenly."""
+        dataset = _create_dataset(num_rows=10, n_unique_natural_ids=3)
+        partitioner = GroupedNaturalIdPartitioner(
+            partition_by="natural_id",
+            group_size=2,
+            mode="strict",
+            sort_unique_ids=self.sort_unique_ids,
+        )
+        partitioner.dataset = dataset
+        with self.assertRaises(ValueError) as context:
+            _ = partitioner.load_partition(0)
+        self.assertIn(
+            "Strict mode requires that the number of unique natural ids is perfectly "
+            "divisible by the group size.",
+            str(context.exception),
+        )
+
+    def test_too_big_group_size(self) -> None:
+        """Test raises if the group size > than the number of unique natural ids."""
+        n_unique_natural_ids = 3
+        dataset = _create_dataset(
+            num_rows=10, n_unique_natural_ids=n_unique_natural_ids
+        )
+        partitioner = GroupedNaturalIdPartitioner(
+            partition_by="natural_id",
+            group_size=n_unique_natural_ids + 1,
+            mode="allow-bigger",
```

Comment:
My understanding is that `allow-smaller` in an scenario like the one tested here. It doesn't work because `num_groups` will be 0. This links to a previous comment in `_create_int_partition_id_to_natural_id()`